framework/src/meshgenerators/TransfiniteMeshGenerator.C

```diff
@@ -0,0 +1,513 @@
+  std::vector<std::string> string_points;
+  MooseUtils::tokenize(parameter, string_points, 1, "\n");
+  if (string_points.size() != np)
+    mooseError("DISCRETE: the number of discrete points does not match the number of points on the"
+               "opposite edge.");
+
+  auto it = 0;
+  for (unsigned int iter = 0; iter < string_points.size(); iter++)
+  {
+    std::vector<Real> point_vals;
+    MooseUtils::tokenizeAndConvert(string_points[iter], point_vals, " ");
+    edge[it] = Point(point_vals[0], point_vals[1], point_vals[2]);
+    it++;
+  }
+
+  return edge;
+}
+
+std::vector<Point>
+TransfiniteMeshGenerator::getCircarcEdge(const Point & P1,
+                                         const Point & P2,
+                                         const std::string & parameter,
+                                         const Point & outward,
+                                         const std::vector<Real> & param_vec)
+{
+  std::vector<Point> edge(param_vec.size());
+
+  std::vector<Real> param_coords;
+  MooseUtils::tokenizeAndConvert(parameter, param_coords, ";");
+  Point P3;
+  if (param_coords.size() == 1)
+  {
+    P3 = computeMidPoint(P1, P2, param_coords[0], outward);
+  }
+  else
+  {
+    P3 = Point(param_coords[0], param_coords[1], param_coords[2]);
+  }
+
+  Real rad = computeRadius(P1, P2, P3);
+  Point P0 = computeOrigin(P1, P2, P3);
+
+  // need to shift to center of coordinates to find the corresponding radians
+  Point x0 = (P1 - P0);
+  Point x1 = (P2 - P0);
+
+  // The case when the edge spans quadrants 1 and 4 requires special treatment
+  // to periodically switch we compute the angle that needs added to one edge
+  // to identify the entire edge span
+  mooseAssert(x0.norm() > 0.0 && x1.norm() > 0.0,
+              "The point provided cannot generate an arc circle on the edge specified");
+
+  Real arclength = std::acos((x0 * x1) / x0.norm() / x1.norm());
+
+  Real a = std::atan2(x0(1), x0(0));
+  Real b = std::atan2(x1(1), x1(0));
+
+  if (a < 0)
+    a = a + 2 * M_PI;
+  if (std::abs(b - a) > M_PI)
+    b = a + arclength;
+
+  auto it = 0;
+  for (auto rx : param_vec)
+  {
+    Real interval = getMapInterval(rx, 0.0, 1.0, a, b);
+    Real x = P0(0) + rad * std::cos(interval);
+    Real y = P0(1) + rad * std::sin(interval);
+    edge[it] = Point(x, y, 0.0);
+    it++;
+  };
+
+  return edge;
+}
+
+Real
+TransfiniteMeshGenerator::getMapInterval(
+    const Real & xab, const Real & a, const Real & b, const Real & c, const Real & d) const
+// this routine maps a point x\in[a, b] to the corresponding point in the interval [c, d]
+{
+  mooseAssert(std::abs(b - a) > 0.0,
+              "The input interval [a, b] is empty, check that a and b are not identical.");
+  Real xcd = c + (d - c) / (b - a) * (xab - a);
+
+  return xcd;
+}
+
+std::vector<Real>
+TransfiniteMeshGenerator::getPointsDistribution(const Real & edge_length,
+                                                const unsigned int & np,
```

Comment:
```suggestion
                                                const unsigned int np,
```

Reply:
I've been curious about this. 
I think both are perfectly equivalent?

Reply:
For an inline function that doesn't call any non-inline functions they're equivalent.  For anything where the compiler has to generate an actual binary function, pass-by-reference passes a pointer on the stack and dereferences it again afterward.  And that gets worse once you start calling other (non-inline) functions from inside your first function.  In that case the pass-by-value code can just keep the value in a register (or at worst on its thread-local stack) for as long as it wants, can cache calculations which used it, etc., but the pass-by-reference code has to take a new dereference every time the parameter appears in a new expression after a function call, because it can't prove that the function call didn't have a side effect which changed the parameter, because `[[pure]]` *still* isn't in a C++ standard...

Of course here it doesn't make a difference whether something takes ten nanoseconds or eleven; I just like having code that serves as a good example for inner kernels elsewhere where it does make a difference.  "Pass anything you don't expect to change by const reference" is an easy wrong heuristic to pick up, specifically because passing something like that by *value* is often a bigger mistake.

Reply:
wow I guess I knew that `const` was just a syntax thing but I never thought it through the implications

I ll try to be more careful

Reply:
Well, const has semantics in some cases, when you use it to pick one overload over another.  But yeah, the fact that you're allowed to lie about it and const_cast around it means the compiler isn't allowed to believe you even when you know you're not lying.

But here the tricky concept isn't "const", it's ... something like "restrict"?  Fortran has had that implicitly forever, C got a keyword for it in C99, but the C++ standards people still can't agree about *exactly* what it should mean.  Colloquially, though, "here we can only get at this data through this one pointer so don't worry about it being changed unexpectedly via any other dereferences or functions" is a nice thing to be able to tell an optimizing compiler.